feat: Inventory swapping behaviour

Koboo opened the pull request:
## Proposed changes

Adds an InventorySwapEvent, which gets dispatched if the Player.openInventory gets called and there was an already opened inventory. The client will "swap" the new inventory with the previous one, which results in not resetting the cursor position within the inventory.

- Issue: https://github.com/Minestom/Minestom/issues/2644
- Discussion: https://github.com/Minestom/Minestom/discussions/2643

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before
merging your code._

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

DasBabyPixel commented on src/main/java/net/minestom/server/entity/Player.java:
```diff
@@ -1745,7 +1750,12 @@ public boolean openInventory(@NotNull Inventory inventory) {
-        return !inventoryOpenEvent.isCancelled();
+        boolean wasSuccessfulOpened = !inventoryOpenEvent.isCancelled();
+        if (wasSuccessfulOpened && previouslyOpenInventory != null) {
+            InventorySwapEvent inventorySwapEvent = new InventorySwapEvent(previouslyOpenInventory, this.openInventory, this);
```
event is cancellable but cancelled not handled

Koboo replied:
Thank you ❤️

DasBabyPixel commented on src/main/java/net/minestom/server/event/inventory/InventorySwapEvent.java:
```diff
@@ -0,0 +1,69 @@
+     *
+     * @return the new inventory
+     */
+    public AbstractInventory getNewInventory() {
```
NotNull

Koboo replied:
❤️

DasBabyPixel commented:
Could also add a test to ensure the InventoryCloseEvent is not called twice. Considering with this PR that would be intended behaviour and is (I think) not tested anywhere

Koboo commented:
I've added a test, to ensure no close event is fired during the `Player.openInventory(..)` call. 😄

mworzala commented:
Currently opening a new inventory does not reset your cursor position which seems to be the major benefit of this PR. So I am assuming I am missing some context? Additionally, you can get the new and old inventory from `InventoryOpenEvent` so this doesn't seem to add anything.

(example included of opening new inventories without the cursor being reset)

https://github.com/user-attachments/assets/e7f4a393-0734-438c-b2aa-5d4b9a468b52

DasBabyPixel commented:
awkward
but agreed
it's a little unintuitive to get the old inventory via 
```
InventoryOpenEvent event = ...
event.getPlayer().getOpenInventory()
```
Thoughts about adding a method to the InventoryOpenEvent something like getOldInventory or getPreviousInventory which wraps that call?

mworzala commented:
Yeah I think a getOldInventory would be reasonable there

Koboo commented:
So if I understand correctly: 
1. Remove `InventorySwapEvent`
2. Add `InventoryOpenEvent.getOldInventory()` wrapper call for `Player.getOpenInventory()`
3. Document the behaviour on `InventoryOpenEvent` and `Player.openInventory(..)` javadocs

I can do that, but I'm in favor of the `InventorySwapEvent` since it's a pretty specific case. For me, a check if `InventoryOpenEvent` is a "swap" or a classic "open" would be pretty annoying and counter intuitive. (_Why would I check for client-side-closing inventories on an `InventoryOpenEvent`?_)

Maybe one more thing we could try:
1. `InventorySwapEvent extends InventoryOpenEvent`
2. Do the check directly in the `openInventory(..)` method 

This way we would still have a specific event, instead of relying on wrapper calls + only the initialization call for the current `InventoryOpenEvent` itself changes. It wouldn't cause any breaking- or behaviour-changes but clearly indicates, that Swapping differs from just Opening.

GoldenStack commented:
I'm still not convinced that InventorySwapEvent is necessary (although I don't really have any authority so I don't know how much this matters). I feel like at the minimum, a few types could be swapped between Inventory and AbstractInventory (like you mentioned) and that would support all of the functionality that you'd need.

Also, why would you have to check for client-side closing inventories in InventoryOpenEvent?

Koboo commented:
I think I wrote it a lilttle bit misleading. The question I mentioned is an example question a user could ask himself. 

If we use the InventoryOpenEvent wrapper method for the old inventory, you would have to check for the client-side swapping behaviour inside the open listener like:
```java
public class InventoryOpenListener extends Consumer<InventoryOpenEvent> {
  public void accept(InventoryOpenEvent event) {
    AbstractInventory inventory = event.getOldInventory();
    if(inventory != null) {
      // Inventory will be swapped client-side
    } 
  }
}
```

I think it's a bit weird to check that in an open event, but as you said, I'm not the decision maker either. I'm here to PR whatever is the cleanest way of doing it. 😅

mworzala commented:
I do see the point about checking on open seeming like an odd place, but im not sure an entirely new event is necessary for this. At least in my experience it's almost never relevant to know if an inventory is being 'swapped' vs close + reopen so delegating to the open event seems ok

Koboo commented:
I've implemented the delegation method in the `InventoryOpenEvent`, added javadocs to the `player.setInventory(Inventory inventory)` method and updated my Tests. :)
